File: app/src/gui/helpers/ImageCrawlWorker.cpp
#include "ImageCrawlWorker.h"
#include "web/ImageCrawler.h" // Assumed C++ equivalent
#include <QDir>
#include <QFileInfo>

ImageCrawlWorker::ImageCrawlWorker(const QVariantMap& config, QObject* parent)
    : QThread(parent), m_config(config), m_crawler(nullptr), m_downloaded(0)
{
}

ImageCrawlWorker::~ImageCrawlWorker()
{
    // Ensure crawler is deleted if it was created
    delete m_crawler;
}

void ImageCrawlWorker::run()
{
    try {
        // Create directories
        QDir().mkpath(m_config["download_dir"].toString());
        QString screenshotDir = m_config["screenshot_dir"].toString();
        if (!screenshotDir.isEmpty()) {
            QDir().mkpath(screenshotDir);
        }
            
        m_crawler = new ImageCrawler(
            m_config["url"].toString(),
            m_config["headless"].toBool(),
            m_config["download_dir"].toString(),
            m_config["browser"].toString(),
            screenshotDir, 
            m_config["skip_first"].toInt(),
            m_config["skip_last"].toInt(),
            nullptr // Create with no parent, will be managed by this thread
        );

        // C++ lambda to replace 'nonlocal'
        auto on_saved = [this](const QString& path) {
            m_downloaded += 1;
            emit status(QString("Saved: %1").arg(QFileInfo(path).fileName()));
        };

        connect(m_crawler, &ImageCrawler::onProgress, this, &ImageCrawlWorker::progress);
        connect(m_crawler, &ImageCrawler::onStatus, this, &ImageCrawlWorker::status);
        connect(m_crawler, &ImageCrawler::onImageSaved, this, on_saved);

        emit status("Starting crawl...");
        m_crawler->run();

        emit finished(m_downloaded, QString("Downloaded %1 image(s)!").arg(m_downloaded));

        delete m_crawler; // Clean up
        m_crawler = nullptr;

    } catch (const std::exception& e) {
        emit error(e.what());
        delete m_crawler; // Clean up on error
        m_crawler = nullptr;
    }
}

File: gui/helpers/image_crawl_worker.py
import os

from PySide6.QtCore import QThread, Signal
from backend.src.web.image_crawler import ImageCrawler


class ImageCrawlWorker(QThread):
    progress = Signal(int, int)      # (current, total)
    status = Signal(str)             # status message
    finished = Signal(int, str)      # (count, message)
    error = Signal(str)              # error message

    def __init__(self, config: dict):
        super().__init__()
        self.config = config

    def run(self):
        try:
            # Create download directory, and screenshot directory if provided
            os.makedirs(self.config["download_dir"], exist_ok=True)
            if self.config.get("screenshot_dir"):
                os.makedirs(self.config["screenshot_dir"], exist_ok=True)
                
            # --- FIX: Passing the entire config dictionary ---
            crawler = ImageCrawler(self.config)
            # --- END FIX ---

            downloaded = 0
            def on_saved(path):
                nonlocal downloaded
                downloaded += 1
                self.status.emit(f"Saved: {os.path.basename(path)}")

            # Note: on_progress signal removed from crawler logic, only using status
            crawler.on_status.connect(self.status.emit)
            crawler.on_image_saved.connect(on_saved)

            self.status.emit("Starting crawl...")
            crawler.run()

            self.finished.emit(downloaded, f"Downloaded {downloaded} image(s)!")

        except Exception as e:
            self.error.emit(str(e))
